### DISCUSSION TITLE
xmake如何引用已经编译好的android so库

### DISCUSSION THREAD

**Discussion by caxiejiang** [xmake-io/xmake#3006] [Q&A]
在大部分示例中，我看到的示例多是如下类型，
target("staticlibrary")
    set_kind("static")
    add_files("src/library/*.c")
target("test")
    set_kind("binary")
    add_files("src/*c")
    add_deps("library")
目前我有一个需求，有几个so库和.a库需要引用，而我是没有源码的，只有库和头文件，我需要把他们引入到我的项目中，在cmakelist或者android.mk中我可以很容易的实现这一块。在xmake.lua中，静态库的添加貌似可以通过add_files把库文件引用进来，但是动态库库的引用目前可以用哪种有效的方式引用进来? 貌似xmake更推荐使用package的方式引用第三方库。
如果xmake支持这种方式，有没有示例可以看下

**Comment by waruqi** [comment]
add_links add_linkdirs

**Reply by caxiejiang** [reply]
非常感謝您的解答，也不好意思因为忙没有回复。其实我之前确实是使用了add_links和add_linkdirs去做了动态库的链接，项目也可以正确编译通过，奇怪的是我把动态库注释掉之后，依然能够编译通过，而我的源工程里面肯定是用到了动态库的一些方法，这让我有些怀疑自己的脚本写的是否正确。另外让我怀疑自己通过xmake生成的动态库的一个原因是，我通过android.mk生成的release版本动态库跟xmake生成的大小不一致(xmake的更小)，差了几百k，我的xmake脚本如下
![image](https://user-images.githubusercontent.com/16971884/199378200-80e8ed4b-9e13-404c-9321-89bdc8dbdde8.png)
我是在windows下通过bat脚本调用xmake的，bat脚本如下:
![image](https://user-images.githubusercontent.com/16971884/199378366-dcbd2090-eb48-4891-823b-cb0169c84046.png)

**Reply by waruqi** [reply]
> 奇怪的是我把动态库注释掉之后，依然能够编译通过，而我的源工程里面肯定是用到了动态库的一些方法，这让我有些怀疑自己的脚本写的是否正确

原本就可以，正常，具体 google 科普下

> 另外让我怀疑自己通过xmake生成的动态库的一个原因是，我通过android.mk生成的release版本动态库跟xmake生成的大小不一致(xmake的更小)，差了几百k

flags 不同，当然可能有变动，-g, -Ox 各种，很正常

**Comment by caxiejiang** [comment]
嗯嗯，多谢解答疑问，我会去进一步了解下编译这块的知识的
